crates/shared/src/trade_finding/oneinch.rs

```diff
@@ -0,0 +1,298 @@
+//! A 1Inch-based trade finder.
+
+use super::{Interaction, Query, Trade, TradeError, TradeFinding};
+use crate::{
+    oneinch_api::{OneInchClient, OneInchError, ProtocolCache, SellOrderQuoteQuery, SwapQuery},
+    price_estimation::gas,
+    solver_utils::Slippage,
+};
+use model::order::OrderKind;
+use primitive_types::H160;
+use std::sync::Arc;
+
+pub struct OneInchTradeFinder {
+    api: Arc<dyn OneInchClient>,
+    disabled_protocols: Vec<String>,
+    protocol_cache: ProtocolCache,
+    referrer_address: Option<H160>,
+}
+
+impl OneInchTradeFinder {
+    async fn quote_and_swap(&self, query: &Query) -> Result<Trade, TradeError> {
+        if query.kind == OrderKind::Buy {
+            return Err(TradeError::UnsupportedOrderType);
+        }
+
+        let allowed_protocols = self
+            .protocol_cache
+            .get_allowed_protocols(&self.disabled_protocols, self.api.as_ref())
+            .await?;
+
+        let (quote, spender, swap) = futures::try_join!(
+            self.api
+                .get_sell_order_quote(SellOrderQuoteQuery::with_default_options(
+                    query.sell_token,
+                    query.buy_token,
+                    allowed_protocols.clone(),
+                    query.in_amount,
+                    self.referrer_address,
+                )),
```

Comment:
Does it make sense to use the `/quote` endpoint at all here? We already have to use `/swap` to get the call data and its response already contains the data we currently pull from the `/quote` endpoint (`gas`, `to_token_amount`).

Reply:
`/swap` responses only contain gas estimates when we don't specify the `disableEstimation` flag - which we need to do since the settlement contract doesn't necessarily have the funds or approvals for doing the swap.

Reply:
Hmm, that's a bummer. Maybe we could suggest to 1Inch to do gas estimates with state overrides? 🤔 
But if we really have to use both endpoints it's safer to pull `to_token_amount` from the `/swap` response, right?

Reply:
So, my reasoning here is:
- We want the `gas_estimate` and `to_token_amount` to come from `/quote` - since this is what we would use if we **don't** simulate the trade.
- When we simulate the trade, we get the executed swap amount (by reading balances before and after in the simulation), so this value isn't super important, the calldata is what is important.

Reply:
Makes sense to me. 👌